torchvision/csrc/cpu/video/Video.cpp

```diff
@@ -0,0 +1,393 @@
+
+// If we are in a Windows environment, we need to define
+// initialization functions for the _custom_ops extension
+// #ifdef _WIN32
+// #if PY_MAJOR_VERSION < 3
+// PyMODINIT_FUNC init_video_reader(void) {
+//   // No need to do anything.
+//   return NULL;
+// }
+// #else
+// PyMODINIT_FUNC PyInit_video_reader(void) {
+//   // No need to do anything.
+//   return NULL;
+// }
+// #endif
+// #endif
+
+const size_t decoderTimeoutMs = 600000;
+const AVSampleFormat defaultAudioSampleFormat = AV_SAMPLE_FMT_FLT;
+// A jitter can be added to the end of the range to avoid conversion/rounding
+// error, small value 100us won't be enough to select the next frame, but enough
+// to compensate rounding error due to the multiple conversions.
+const size_t timeBaseJitterUs = 100;
+
+// returns number of written bytes
+template <typename T>
+size_t fillTensorList(
+    DecoderOutputMessage& msgs,
+    torch::Tensor& frame,
+    torch::Tensor& framePts) {
+  // set up PTS data
+  const auto& msg = msgs;
+
+  float* framePtsData = framePts.data_ptr<float>();
+
+  float pts_s = float(float(msg.header.pts) * 1e-6);
+  framePtsData[0] = pts_s;
+
+  T* frameData = frame.numel() > 0 ? frame.data_ptr<T>() : nullptr;
+
+  if (frameData) {
+    auto sizeInBytes = msg.payload->length();
+    memcpy(frameData, msg.payload->data(), sizeInBytes);
+  }
+  return sizeof(T);
+}
+
+size_t fillVideoTensor(
+    DecoderOutputMessage& msgs,
+    torch::Tensor& videoFrame,
+    torch::Tensor& videoFramePts) {
+  return fillTensorList<uint8_t>(msgs, videoFrame, videoFramePts);
+}
+
+size_t fillAudioTensor(
+    DecoderOutputMessage& msgs,
+    torch::Tensor& audioFrame,
+    torch::Tensor& audioFramePts) {
+  return fillTensorList<float>(msgs, audioFrame, audioFramePts);
+}
+
+std::string parse_type_to_string(const std::string& stream_string) {
+  static const std::array<std::pair<std::string, MediaType>, 4> types = {{
+      {"video", TYPE_VIDEO},
+      {"audio", TYPE_AUDIO},
+      {"subtitle", TYPE_SUBTITLE},
+      {"cc", TYPE_CC},
+  }};
+  auto device = std::find_if(
+      types.begin(),
+      types.end(),
+      [stream_string](const std::pair<std::string, MediaType>& p) {
+        return p.first == stream_string;
+      });
+  if (device != types.end()) {
+    return device->first;
+  }
+  AT_ERROR("Expected one of [audio, video, subtitle, cc] ", stream_string);
+}
+
+MediaType parse_type_to_mt(const std::string& stream_string) {
+  static const std::array<std::pair<std::string, MediaType>, 4> types = {{
+      {"video", TYPE_VIDEO},
+      {"audio", TYPE_AUDIO},
+      {"subtitle", TYPE_SUBTITLE},
+      {"cc", TYPE_CC},
+  }};
+  auto device = std::find_if(
+      types.begin(),
+      types.end(),
+      [stream_string](const std::pair<std::string, MediaType>& p) {
+        return p.first == stream_string;
+      });
+  if (device != types.end()) {
+    return device->second;
+  }
+  AT_ERROR("Expected one of [audio, video, subtitle, cc] ", stream_string);
+}
```

Comment:
Those two functions seem almost exactly the same, should we try to make them share code?